packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java

```diff
@@ -0,0 +1,222 @@
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gms.maps.model.Marker;
+import com.google.android.gms.maps.model.MarkerOptions;
+import com.google.maps.android.clustering.Cluster;
+import com.google.maps.android.clustering.ClusterItem;
+import com.google.maps.android.clustering.ClusterManager;
+import com.google.maps.android.clustering.view.DefaultClusterRenderer;
+import com.google.maps.android.collections.MarkerManager;
+import io.flutter.plugin.common.MethodChannel;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Controls cluster managers and exposes interfaces for adding and removing clusteritems for
+ * specific cluster managers.
+ */
+class ClusterManagersController implements GoogleMap.OnCameraIdleListener, ClusterListener {
+  @NonNull private final Context context;
+  @NonNull private final HashMap<String, ClusterManager<MarkerBuilder>> clusterManagerIdToManager;
+  @NonNull private final MethodChannel methodChannel;
+  @Nullable private MarkerManager markerManager;
+  @Nullable private GoogleMap googleMap;
+  @Nullable private ClusterItemListener clusterItemListener;
+
+  ClusterManagersController(MethodChannel methodChannel, Context context) {
+    this.clusterManagerIdToManager = new HashMap<>();
+    this.context = context;
+    this.methodChannel = methodChannel;
+  }
+
+  void init(GoogleMap googleMap, MarkerManager markerManager) {
+    this.markerManager = markerManager;
+    this.googleMap = googleMap;
+  }
+
+  void setClusterItemListener(@Nullable ClusterItemListener listener) {
+    clusterItemListener = listener;
+    initListenersForClusterManagers(this, listener);
+  }
+
+  private void initListenersForClusterManagers(
+      @Nullable ClusterListener clusterListener,
+      @Nullable ClusterItemListener clusterItemListener) {
+    for (Map.Entry<String, ClusterManager<MarkerBuilder>> entry :
+        clusterManagerIdToManager.entrySet()) {
+      initListenersForClusterManager(entry.getValue(), clusterListener, clusterItemListener);
+    }
+  }
+
+  private void initListenersForClusterManager(
+      ClusterManager<MarkerBuilder> clusterManager,
+      @Nullable ClusterListener clusterListener,
+      @Nullable ClusterItemListener clusterItemListener) {
+    clusterManager.setOnClusterClickListener(clusterListener);
+    clusterManager.setOnClusterItemClickListener(clusterItemListener);
+  }
+
+  /** Adds new ClusterManagers. */
+  void addClusterManagers(List<Object> clusterManagersToAdd) {
+    if (clusterManagersToAdd != null) {
+      for (Object clusterToAdd : clusterManagersToAdd) {
+        addClusterManager(clusterToAdd);
+      }
+    }
+  }
+
+  /** Adds new ClusterManager. */
+  void addClusterManager(Object clusterManagerData) {
+    String clusterManagerId = getClusterManagerId(clusterManagerData);
+    if (clusterManagerId == null) {
+      throw new IllegalArgumentException("clusterManagerId was null");
```

Comment:
How would you like me to change this?
Exceptions thrown in the native code are converted to PlatformExceptions by the method channel and this is handled as an PlatformException at dart code.

Reply:
IllegalArgumentExceptions are used in the Android platform on similar places.
Exceptions thrown in the method channel method handlers are converted to PlatformExceptions by the method channel code:

E/flutter (12022): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: **PlatformException**(error, clusterManagerId was null, null, java.lang.IllegalArgumentException: clusterManagerId was null

As clusterManagerUpdates should be created by the [ClusterManagerUpdates](https://github.com/flutter/packages/blob/dce6f0c91fcf01ee3a67e3475155ec9f16ce2a66/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager_updates.dart#L11) class this error should not be never seen by the developers.

How would you like me to change this handling?

Reply:
@reidbaker @stuartmorgan do you still need changes to this?

Reply:
I'm fine with this staying as is for now since a) it should never actually happen, and b) it will go away when we convert to Pigeon, which I plan to do this quarter.